### ISSUE TITLE
[Feature] Simulator for Edge-cloud Collaborative AI: a Lifelong Learning example

### ISSUE THREAD

**Issue by MooreZheng** [kubeedge/ianvs#348]
**Background**:
For most distributed AI scheme developers, it is often complicated and a heavy burden to establish and deploy a large-scale cloud-edge collaborative system. Currently, ianvs provide a algorithm tester in one node, and accuracy-wise metrics can be tested through test datasets. However, the system-wise metrics then become impossible to measure in a real-world manner for large-scale nodes and out-of-the box testing frequently. 

This issue then aims to provide an industrial-grade distributed collaborative system simulation for a large-scale deployment of an edge-cloud collaborative AI system. To show how it works, this issue selects an edge-cloud collaborative lifelong learning example for testing. 

**What contents are to be added/modified**:
In the test case controller of Ianvs, it is expected to add a simulation controller. It will provide Ianvs with a worker-in-worker system. Then even if one is using a signle machine, simulated workers can be generated to simulated multi-node systems. 

The modules in the simulation controller could be as follows:
- A Simulation Environment Administrator is used to
1. Parse the system config to establish the simulation
2. Check the host environment, e.g., check if the host has all requirements met, and whether memory > 4GB
3. Build the simulation environment
4. Create and deploy the modules needed in the simulation environment
5. Close and delete the simulation environment
- The Simulation Job Administrator is critical to manage the simulation job, and provides the following functions:
1. Build the images of algorithms to be tested, e.g., docker image
2. Generate the YAML file of the simulation job
3. Deploy and delete the simulation job with workers
4. List-watch the results of the simulation job with workers

==============
Appendix on Feb 11, 2022
The KubeEdge SIG AI is chartered to facilitate Edge AI applications with KubeEdge. An overview of SIG AI activities can be found in this [charter](https://github.com/kubeedge/community/blob/master/sig-ai/charter.md). 

KubeEdge-Sedna has released the first [edge-cloud collaborative lifelong learning](https://github.com/kubeedge/sedna/releases/tag/v0.3.0) on June 2021, together with [a hands-on example](https://sedna.readthedocs.io/en/latest/examples/lifelong_learning/atcii/README.html) and [a free playground on Katacoda](https://www.katacoda.com/kubeedge-sedna/scenarios/lifelong-learning-example). The sceme learn coming tasks from the edge based on previously learned tasks on the cloud knowledgebase. 

**Why is this project needed**:
According to the survey on edge AI landing challenges, the No.1 advice goes to "providing public datasets, pre-processing, and baseline codes to build benchmarks". This advice gets 82.18%, 92.98%, 87.10%, and 86.67%  votes among all participants, the industial, acadamic and student community on edge ai, respectively. 

Our Effort starts from the edge-cloud collaborative benchmarks. This project focuses on measuring and validating the desired behaviors for an epoch-making Edge AI scheme, i.e., [Edge-cloud Collaborative Lifelong Learning](https://sedna.readthedocs.io/en/latest/proposals/lifelong-learning.html). 

**What contents are to be added/modified**:
This project will help all Edge AI application developers to validate and select the best-matched algorithm for lifelong learning. Parts of the effort are developing test cases on the existing scheme of Edge-cloud Collaborative Lifelong Learning on KubeEdge-Sedna, including interfaces for benchmark specification, algorithms like preprocessing and metrics, and even baselines. 

**Other information**: 
Recommended Skills: TensorFlow/Pytorch, Python
[Sedna lifelong-learning introduction](https://segmentfault.com/a/1190000040132422/en) 
[Sedna guide](https://sedna.readthedocs.io/en/latest/index/guide.html)
[Sedna lifelong-learning proposal](https://sedna.readthedocs.io/en/latest/proposals/lifelong-learning.html)
[How to contribute to Sedna](https://sedna.readthedocs.io/en/latest/contributing/prepare-environment.html)

**Related issue for reference (if any)**:
[#58 Add an example of implementation of Federated Learning to ReID](https://github.com/kubeedge/sedna/issues/58)
[#96 Edge AI Benchmark review](https://github.com/kubeedge/sedna/issues/96)
[#118 Edge AI Benchmark: Consolidate/Prioritize metrics](https://github.com/kubeedge/sedna/issues/118)
[#119 Edge AI Benchmark: add step-by-step instructions for users to benchmark their works](https://github.com/kubeedge/sedna/issues/119)
[#120 Edge AI Benchmark: add leaderboard/forum](https://github.com/kubeedge/sedna/issues/120)
[#274 KubeEdge SIG AI: Benchmarks for Edge-cloud Joint Inference](https://github.com/kubeedge/sedna/issues/274) 

------------
Several breath-taking Edge-AI scenarios have been prepared for benchmarking: looking forward to seeing your codes involved in real-world robots, outer-space satellites, and industrial production lines!

**Comment by MooreZheng** [comment]
KubeEdge has launched a new sub-project [KubeEdge-Ianvs](https://github.com/kubeedge/ianvs) for distributed collaborative AI benchmarking to handle this issue.

@iszhyang starts a great ospp project working on a simulator for this issue! Pls refer to: 
- Proposal Pull Request: https://github.com/kubeedge/ianvs/pull/35
- Code Pull Request: https://github.com/kubeedge/ianvs/pull/39
- Demo Video（in Chinese）on KEAW'22: https://www.bilibili.com/video/BV1nD4y1Y7cJ

A related issue on cloud robotics is available at KubeEdge-Ianvs: https://github.com/kubeedge/ianvs/issues/48

**Comment by MooreZheng** [comment]
See simulator related issue, PR and demo:
- Issue: https://github.com/kubeedge/ianvs/issues/348
- Proposal Pull Request: https://github.com/kubeedge/ianvs/pull/35
- Code Pull Request: https://github.com/kubeedge/ianvs/pull/39
- Demo Video（in Chinese）on KEAW'22: https://www.bilibili.com/video/BV1nD4y1Y7cJ

**Comment by MooreZheng** [comment]
## LFX Mentorship 2026 Term 3 Project

### KubeEdge-Ianvs Simulation Sandbox: Environment-Isolated Execution and System Metrics Profiling

**Description：**
For most distributed AI scheme developers, establishing and deploying a large-scale cloud-edge collaborative system is often complicated and burdensome. While KubeEdge-Ianvs currently provides a single-node algorithm tester to evaluate accuracy-wise metrics using test datasets, measuring system-wise metrics, such as bandwidth, computing power, and peak memory, in a real-world manner for large-scale nodes is extremely difficult and costly. Furthermore, executing all test cases within a single shared Python process easily triggers dependency conflicts, path contamination, and fatal OOM crashes when heavy-load models like LLMs, VLAs, and foundation models run alongside lightweight examples. To address these challenges, this project would like to introduce an industrial-grade distributed collaborative system simulation using a worker-in-worker approach on a single machine, providing low-cost, scalable testing capabilities, robust environment isolation, and precise system-level metric profiling. 

**Expected Outcome:**

1. Restoration of Simulation Core Functions: Considering previous proposal Restore and extend the Ianvs 2022 simulation proposal (Ianvs PR #35) and implementation (Ianvs PR #39, with 5+ identified breakages), to build the Simulation Controller to isolate each test case within an independent transient runtime environment, and enforce system-level resource quotas and boundary control mechanisms to restrict edge node resources (CPU and memory), avoiding dependency conflicts and OOM risks. Key components include: 
  - Environment Administrator of Simulation Controller: Introduce a simulation controller within the test case controller to provide a worker-in-worker system on a single machine, simulating multi-node systems. Implement the Simulation Environment Administrator to parse system configurations, check host environment requirements (e.g., memory > 4GB), and automatically build, deploy, close, and delete the simulation environment.
  - Simulation Job Administrator of Simulation Controller: Develop the critical Simulation Job Administrator to handle algorithm image building (e.g., Docker), YAML generation, job deployment/deletion, and list-watching of simulation results with workers. Simultaneously, deploy an isolated execution layer using transient runtime environments and system resource quotas (CPU and memory) to completely prevent dependency conflicts and OOM crashes.
2. Verify Cluster with Multi-Dimensional Metrics: Complete KubeEdge-native cluster simulation validation using kind + edgecore + Sedna all-in-one scripts.
3. Multi-Dimensional Metrics Integration: Align underlying system metrics (CPU utilization, peak memory, wall-clock time) with upper-layer algorithm metrics, presenting them uniformly in the existing StoryManager leaderboard to achieve end-to-end comprehensive performance evaluation for distributed AI. 

**Recommended Skills:** KubeEdge-Ianvs, KubeEdge-Sedna, KubeEdge, Kubernetes, Docker, Linux Kernel mechanisms, Go, Python, Benchmark, AI/ML

**Useful links:** 
[OSPP] Proposal for Edge Intelligence Benchmark for Edge-Cloud Collaborative Lifelong Detection https://github.com/kubeedge/ianvs/pull/35
[OSPP] The codes of Edge Intelligence Benchmark for Edge-Cloud Collaborative Lifelong Detection https://github.com/kubeedge/ianvs/pull/39

**Comment by RudraDudhat2509** [comment]
Hi @MooreZheng, instead of just reading the proposal I went and actually ran the old 2022 code this is meant to restore (PR #35 / #39), and found 2 real breakages in the host-env check step (`simulation_system_admin.py`), both with the same root cause basically.

**`check_host_docker()` reports success even when Docker is unreachable.** It runs `"docker version | head -n 2"` with `subprocess.run(..., check=True)`. Since it's piped through `head`, the exit code `check=True` is actually looking at is `head`'s, not `docker version`'s. `head` succeeds as long as it gets any stdout, regardless of whether the daemon actually responded. Reproduced this on my machine with Docker installed but the daemon not running:

```
error during connect: ...dockerDesktopLinuxEngine...
[INFO] check docker successful
```

It logs "successful" and just returns like nothing happened. The `if check_docker.returncode != 0:` block right after (which should trigger the auto-install fallback) can never actually run because of this.

**`check_host_kind()` crashes instead of running its own auto-install fallback.** Same root issue, opposite result: `"kind version"` isn't piped through anything, so when kind isn't installed, `check=True` raises immediately:

```
CalledProcessError: Command 'kind version' returned non-zero exit status 1.
```

So the `else:` block right below it (supposed to curl-install kind automatically) never gets a chance to run either — dead code, same as the docker case just flipped (one hides the failure, the other never gets to self-heal from it).

Small side note while I was in there: `build_simulation_enviroment` / `destory_simulation_enviroment` has a typo (destory -> destroy), not a real bug, just figured I'd mention it since I'll probably be touching these functions anyway.

Before I put together an actual proposal - both of these come down to leaning on `subprocess.run(shell=True, check=True)` against raw CLI output. For the new Environment Administrator, is the plan to keep shelling out to the docker/kind binaries (just fixing the exit code checks, e.g. dropping the head pipe), or would you rather move to something like the Docker SDK / structured output instead, given this is basically the make-or-break step before any test case even starts. Lmk what direction you had in mind.

**Comment by DsThakurRawat** [comment]
Hi @MooreZheng — I ran the 2022 host-environment step (PR #35 / #39) on a current machine to find the "5+ identified breakages" the Term 3 description mentions.

@RudraDudhat2509 already covered the two `subprocess.run(..., check=True)` problems in `check_host_docker()` / `check_host_kind()` above, so I won't repeat those. Everything below is what I found *after* those two, on the same file.

**1. `get_host_number_of_cpus()` crashes on modern `util-linux`.** The grep pattern is `lscpu | grep CPU:`, but current `lscpu` emits the label as `CPU(s):` — there is no line matching `CPU:` at all, so the pipeline produces empty output and the parse raises:

```
$ lscpu | grep CPU:
$ echo $?
1
IndexError: list index out of range
```

On any host that clears the memory gate, this is where the run dies. `check_host_enviroment()` calls docker → kind → memory → cpu in that order, so even with the docker and kind checks fixed, `build_simulation_enviroment()` still stops here before a single test case starts.

There are two things wrong at once. Grepping `CPU(s):` unanchored also matches `NUMA node0 CPU(s):`, whose value is a *range* (`0-11`), not a count — so an unanchored fix silently swaps a core count for a range. And the label is locale-dependent; under a non-English `LC_ALL` it isn't `CPU(s)` at all. What worked for me was `LC_ALL=C lscpu | grep '^CPU(s):'`.

**2. Both parsers are reading the `repr()` of a `bytes` object rather than decoded text.** `stdout.read()` returns `bytes`, and the code does `str(bytes)`, which yields the literal string `"b'MemFree:  1881580 kB\n'"` — quotes, `b` prefix and a two-character `\n` included. The memory parser gets the right number by luck (`split(" ")[0]` happens to cut before the trailing junk). The CPU parser doesn't: its `.split("\\")[0]` is there purely to strip the backslash of that escaped `\n`, which is the fingerprint of the same bug. Decoding the stream removes the need for that hack entirely.

**3. The sedna installer URL disagrees between build and teardown.** `build_simulation_enviroment()` fetches `.../kubeedge/sedna/master/scripts/installation/all-in-one.sh`, while `destory_simulation_enviroment()` twelve lines below fetches the same script from `.../main/...`. Both currently return 200 with byte-identical content (`md5 a4fb9ed624cd6b22e6ccc9729513f4bf`), so this is not breaking today — but `main` is sedna's default branch and `master` is a stale alias, so the two halves of the lifecycle are one branch deletion away from disagreeing about which script they run.

**4. `destory_simulation_enviroment()` is exported but never called.** It's re-exported in `simulation_system_admin/__init__.py`, but the only consumer, `core/cmd/obj/benchmarkingjob.py`, imports `build_simulation_enviroment` alone and calls it in `run()` with no teardown counterpart. So item 5 of this issue's spec — "Close and delete the simulation environment" — exists as code but is unreachable: every benchmarking run leaks its kind cluster. That seems directly relevant to the environment-isolation goal in the Term 3 description, since isolation is only worth as much as the cleanup that follows it.

I've opened #656 with the mechanical half of this — items 1–3 plus the two `check=` fixes, so the file isn't left half-repaired. Item 4 I left alone deliberately: wiring teardown in means deciding whether it's a `finally` in `BenchmarkingJob.run()`, a context manager, or something the new Environment Administrator owns, and that's your call rather than mine.

One question on direction, since it shapes the rest. `check_host_memory()` gates on `MemFree` from `/proc/meminfo`, which on a warm Linux host is mostly an artifact of page cache — my box reports 0.80 GB free against 3.93 GB available. So a machine with nearly 4 GB genuinely available is rejected by a check whose stated intent is "≥4GB". Should the restored administrator keep reading `/proc/meminfo` and just move to `MemAvailable`, or is the intent to drop the raw-CLI parsing here altogether — `os.cpu_count()` and `psutil`/`os.sysconf` for the host probes, Docker SDK for the daemon — given that every breakage in this file traces back to parsing human-readable CLI output?

**Comment by Lu2077** [comment]
"Hi @ZimuZheng and mentors! I have just applied to this mentorship program through the LFX Dashboard. I'm very excited about the Edge-cloud Collaborative AI Simulator project, especially since I have experience with Python and I'm highly motivated to learn and integrate it with Go in the KubeEdge-Ianvs ecosystem. I'm looking forward to the review process and contributing to the community!"

**Comment by MooreZheng** [comment]
# Pre-test for LFX Mentorship 2026 Term 3

CNCF — KubeEdge: Ianvs Simulation Sandbox: Environment-Isolated Execution and System Metrics Profiling (2026 Term 3)

---

## Introduction

Ianvs is a distributed synergy AI benchmarking framework for edge-cloud collaborative algorithms under KubeEdge. This pre-test evaluates candidates for the LFX Mentorship 2026 Term 3 Simulation Sandbox position through two mandatory tasks and one bonus task, assessing environment competency, codebase debugging ability, and isolation design capability.

---

## Submission Requirements

Each candidate must create **ONE pre-test Discussion** within the `Show and tell` category (https://github.com/kubeedge/ianvs/discussions/categories/show-and-tell). Name it `LFX 2026 Term 3 Simulation Sandbox: analysis by xxx (candidate name)`. **Each task must be submitted as one separate Comment within the same Discussion.**

---

## Task 1: Working Environment Validation

Ianvs is a distributed synergy AI benchmarking framework for edge-cloud collaborative algorithms, developed under KubeEdge. The `pcb-aoi` single-task learning benchmark for fault detection serves as the validation entry point. Validate a working Ianvs environment and demonstrate end-to-end benchmark execution—prerequisite for all subsequent simulation sandbox work.

### Actions

Task 1 requires the following actions, where every command executed and its output must be recorded. 
1. **Set up**
   1.1. Conduction the following actions: 1) Clone the repository: `git clone https://github.com/kubeedge/ianvs.git`; 2) Install Python dependencies in a virtual environment: `pip install -r requirements.txt`; 3) Verify the installation: `ianvs --help` or equivalent entry point.
   1.2. Record every command executed, in order
   1.3. Document any issues encountered (dependency conflicts, version mismatches, missing packages) and how they were resolved
   1.4. Capture the full terminal output (stdout and stderr) as evidence


2. **Run example**
   2.1. Navigate to `examples/pcb-aoi/singletask_learning_bench/fault_detection`
   2.2. Execute the benchmark following the example README
   2.3. Confirm the benchmark completes successfully and produces test results
   2.4. Provide the complete terminal transcript and a video showing setup and successful run
   2.5. If the run fails, document the failure in detail and describe troubleshooting steps taken 

In the discussion, please provide one Comment containing the full terminal transcript and any issue-resolution notes.

### Scoring

| Task | Item | Score |
|------|------|-------|
| Task 1 | Set up | 10 |
| Task 1 | Run Example | 15 |
| **Task 1** | **Total** | **25** |

---

## Task 2: Simulation Codebase Bug Identification

Identify at least 3 specific bugs or breakages in the simulation codebase and propose concrete corrections—validating the candidate's ability to read, understand, and debug the simulation system code that the project aims to restore and extend.

The file `core/testcasecontroller/simulation_system_admin/simulation_system_admin.py` implements the Simulation System Administrator—responsible for parsing system configurations, checking host environment requirements, and building/deploying/closing/deleting the simulation environment.

### Actions

1. **Understanding**
   1.1. Thoroughly read and write down the understanding of `core/testcasecontroller/simulation_system_admin/simulation_system_admin.py`
   1.2. Understand and write down the role of each function: `check_host_docker()`, `check_host_kind()`, `get_host_free_memory_size()`, `check_host_memory()`, `get_host_number_of_cpus()`, `check_host_cpu()`, `check_host_enviroment()`, `build_simulation_enviroment()`, `destory_simulation_enviroment()`
   1.3. Point out the difference between the proposal and the implementation. Cross-reference with the [simulation proposal](https://github.com/kubeedge/ianvs/blob/main/docs/proposals/simulation/simulation.md) and the [sandbox-engine proposal](https://github.com/kubeedge/ianvs/blob/main/docs/proposals/simulation/sandbox-engine/sandbox-engine.md).


2. **Debug**
   2.1. For each bug, provide: exact line number and explanation;
   2.2. State the impact of each bug on the simulation environment: (a) what functionality breaks, (b) what callers are affected, (c) what deployment scenarios fail;
   2.3. proposed corrected code as a concrete diff or replacement snippet;
   2.4. Bonus: identify more than 3 bugs.

3. **Submit**
   3.1. Create a GitHub Discussion Comment;
   3.2. Include all identified bugs with line numbers, explanations, and proposed fixes.


### Scoring

| Task | Item | Score |
|------|------|-------|
| Task 2 | Understanding | 10 |
| Task 2 | Debug | 30 |
| Task 2 | Debug Bonus | 5 |
| Task 2 | Submit | 5 |
| **Task 2** | **Total** | **50** |

---

## Task 3: Container Isolation Capability Demonstration

Demonstrate the core concept of the Simulation Sandbox—launching an algorithm in an isolated Docker container, capturing execution metrics, and handling OOM gracefully—validating the candidate's understanding of container-based isolation and production-oriented code design.

The prototype directly maps to the **Simulation Job Administrator** component described in the simulation proposal, which must "handle algorithm image building (e.g., Docker), job deployment/deletion, and list-watching of simulation results with workers" while "deploying an isolated execution layer using transient runtime environments and system resource quotas."

Docker exit code `137` = 128 + 9 (SIGKILL signal), indicating the container was killed by the kernel's OOM killer. The prototype must handle this gracefully by returning a structured error dictionary instead of crashing, enabling the caller to distinguish OOM from other failures and adjust resource limits.

### Evidence

The script must be runnable without modification (aside from having Docker installed). **Valid evidence** means: (1) the complete source code, (2) a markdown section with usage examples, and (3) sample output demonstrating both successful execution and OOM handling. Impact: without runnable evidence, the isolation capability cannot be verified.

### Actions

1. **Launch**
   Add a section in the discussion comment which includes the screen shots and video recording the following action: 
   1.1. Accept inputs via CLI arguments or a config dict: `image` (Docker image name/tag), `command` (algorithm command), `cpu_limit` (optional, e.g., `"1.0"`), `memory_limit` (optional, e.g., `"512m"`), `timeout` (optional, wall-clock seconds)
   1.2. Launch the container using `docker run` (via `subprocess` or `docker` Python SDK)
   1.3. Measure wall-clock time from container start to finish
   1.4. Capture stdout and stderr from the container
   1.5. Capture the container exit code

2. **Handle OOM**
   Add a section in the discussion comment which includes the screen shots and video recording the following code action: 
   2.1. If exit code is `137` (OOM Killed), return a structured error dict:
       ```python
       {
           "status": "oom_killed",
           "exit_code": 137,
           "wall_time_s": <measured_time>,
           "stdout": <captured_stdout>,
           "error": "Container killed due to OOM (exit code 137). Consider increasing memory_limit."
       }
       ```
   2.2. For other non-zero exit codes, return a similar structured dict with `"status": "failed"`
   2.3. For exit code 0, return a success dict with `"status": "success"`

3. **README**
   Add a section in the discussion comment which 
   3.1. Explain the purpose, usage, and design decisions
   3.2. Include example commands showing how to run the script
   3.3. Document the structured output format


### Scoring

| Task | Item | Score |
|------|------|-------|
| Task 3 | Launch | 12 |
| Task 3 | Handle OOM  | 8 |
| Task 3 | README | 5 |
| **Task 3** | **Total (bonus)** | **25** |

---

## Scoring Rules

- Only ONE submission counts per candidate per task; if multiple are submitted, only the latest one counts.
- To avoid placeholders, for those whose contents from the GitHub discussion and email are different, the email version will be used for pre-test evaluation.
- Timing of email submission is important. If the same problem and solution have been pointed out by others before (including automated review bots and reviewers), the score will be ZERO without any notification. This applies to both Task 2 and Task 3.
- In a task, if the solution is different from the same problem proposed by others, the candidate can still have partial scores. But please do make the uniqueness clear in the bug report's explanation and impact analysis, otherwise it might result in a score less than 20%.
- Each item will be scored based on completion and quality, and the sum of the scores will be the total score for this task.
- For links that cannot be opened successfully, the score will be ZERO without any notification. Ensure all links are accessible before submission.

| Task | Total |
|------|-------|
| Task 1 | 25 |
| Task 2 | 50 |
| Task 3 | 25 |
| **Grand Total** | **100** |

---

## How to submit

After completing the GitHub submission, compile the **full content of all tasks** into a **`.pdf` or `.docx`** file, and attach: (1) the Discussion link; (2) each task's Comment links. Send all materials via email to **zimu.zheng@huawei.com** and **sjhu21@m.fudan.edu.cn**. All links will be published after the submission deadline under the corresponding tracking issue.

> ⚠️ **Dual-channel submission is mandatory**: both the GitHub Discussion (with per-task Comments) and the email (with compiled PDF/DOCX + links) must be completed. Missing either channel will result in the submission not being evaluated.

---

## Deadline

According to the [official schedule](https://github.com/cncf/mentoring/tree/main/programs/lfx-mentorship/2026/03-Sep-Nov) of the LFX Mentorship 2026 Term 3:

Please complete this pre-test and send all materials via email by **2026-08-28 23:59 UTC (2026-08-29 07:59 UTC+8)**. 

> ⚠️ Both the GitHub Discussion and the email submission must be completed before the deadline. Late submissions on either channel will not be evaluated.

**Comment by nitbig** [comment]
# LFX 2026 Term 3 Simulation Sandbox : Pre-Test Submission

Hello KubeEdge Ianvs maintainers,

I am submitting my pre-test work for the **LFX Mentorship 2026 Term 3 : Ianvs Simulation Sandbox: Environment-Isolated Execution and System Metrics Profiling**.

This Discussion contains my work for the required pre-test tasks. Each task is provided as a separate comment as requested in the pre-test instructions.

 Task 1:  Working Environment Validation

Video Link: [Task 1](https://drive.google.com/file/d/1cswxa1aGs7fyMYiiPy6n-hpna2YEha-t/view?usp=drivesdk)

  - Ianvs environment setup
  - Dependency installation and verification
  - PCB-AOI benchmark execution
  - Terminal transcript and execution evidence

**Comment by nitbig** [comment]
# Task 2: Simulation Codebase Bug Identification

## 1. Understanding

The file under review is:

`core/testcasecontroller/simulation_system_admin/simulation_system_admin.py`

The Simulation System Administrator is responsible for checking the host environment, building the simulation environment, and destroying it after execution.

The current implementation provides the following functions:

`check_host_docker()`

Checks whether Docker is available.

`check_host_kind()`

Checks whether Kind is available.

`get_host_free_memory_size()`

Reads the free memory reported by `/proc/meminfo`.

`check_host_memory()`

Checks whether at least 4 GB of memory is available.

`get_host_number_of_cpus()`

Reads the CPU count using `lscpu`.

`check_host_cpu()`

Checks whether at least 4 CPUs are available.

`check_host_enviroment()`

Runs the Docker, Kind, memory, and CPU checks.

`build_simulation_enviroment()`

Checks the host and then downloads and executes the Sedna all in one installation script.

`destory_simulation_enviroment()`

Downloads and executes the Sedna cleanup script.

The original simulation proposal describes the Simulation System Administrator as responsible for checking the host, building the simulation environment, creating the required modules, and closing and deleting the environment. The sandbox proposal continues to use this component as the foundation for the cluster simulation layer.

## 2. Proposal Comparison

The implementation provides the basic host validation and environment lifecycle functions required by the simulation design.

The main reliability issue I observed is in the way the environment lifecycle depends on external commands and remote scripts. The build and cleanup paths do not independently verify the successful retrieval of the remote installation script, and neither path places a bound on how long the external operation may run.

The sandbox proposal requires environment provisioning to be a reliable foundation for isolated execution. A failure during provisioning or cleanup should therefore produce a clear and bounded result rather than leaving the caller waiting indefinitely or reporting an incorrect lifecycle result.

## 3. Bug Findings

### Bug 1: Environment build does not reliably propagate failure from the Sedna script download

**File:** `core/testcasecontroller/simulation_system_admin/simulation_system_admin.py`

**Function:** `build_simulation_enviroment()`

**Lines:** 155 to 166

**Problem:**

The environment is created through a shell pipeline:

```python
shell_cmd = "curl https://raw.githubusercontent.com/kubeedge/sedna\
/master/scripts/installation/all-in-one.sh | " \
    f"NUM_CLOUD_WORKER_NODES={simulation.cloud_number} " \
    f"NUM_EDGE_NODES={simulation.edge_number} " \
    f"KUBEEDGE_VERSION={simulation.kubeedge_version} " \
    f"SEDNA_VERSION={simulation.sedna_version} " \
    f"CLUSTER_NAME={simulation.cluster_name} bash -"

build_simulation_env_ret = subprocess.run(
    shell_cmd, shell=True, check=True)

**Comment by PragnyaKhandelwal** [comment]
# **LFX Mentorship 2026 Term 3 — Simulation Sandbox Progress**

Mentee: @PragnyaKhandelwal
Mentors: @MooreZheng @hsj576
Proposal: #526 (merged Aug 20, 2026)
Term: September – November 2026

## **Phase 1 — Restoration**
- [ ] Audit open simulation PRs (#656, #698, #745, #820, #896)
- [ ] Fix `simulation_system_admin.py` breakages
- [ ] Wire teardown lifecycle in `benchmarkingjob.py`
- [ ] Verify `build_simulation_environment()` deploys successfully

## **Phase 2 — Simulation Job Administrator (main deliverable)**
- [ ] Docker image building for algorithms under test
- [ ] Job YAML generation from `testenv.yaml` + `algorithm.yaml`
- [ ] Cluster deployment via `kubectl`
- [ ] ConfigMap list-watch for results
- [ ] `SimulationController.run_sandboxed()` implementation

## **Phase 3 — System Metrics + PoC + Docs**
- [ ] System Metrics Profiler (CPU, peak memory, wall-clock)
- [ ] Metrics integration with StoryManager leaderboard
- [ ] PoC: pcb-aoi/incremental_learning_bench
- [ ] PoC: cloud-edge-collaborative-inference-for-llm
- [ ] User guide and configuration reference

#### Related: #526 #495 #496 #497 #499 #500 #908
